src/thunderbots/software/ai/world/ball.cpp

```diff
@@ -122,6 +124,19 @@ void Ball::addStateToBallHistory(const Point &position, const Vector &velocity,
     last_update_timestamps.push_front(timestamp);
 }
 
+int Ball::getHistoryIndexFromTimestamp(Timestamp &timestamp) const
+{
+    std::vector<Timestamp> timestamp_history = getPreviousTimestamps();
+    for (int i = 0; i < timestamp_history.size(); i++)
+    {
+        double timestamp_diff =
+            fabs((timestamp - timestamp_history[i]).getMilliseconds());
+        if (timestamp_diff < POSSESSION_TIMESTAMP_TOLERANCE_IN_MILLISECONDS)
+            return i;  // If timestamp is close to desired timestamp, return the index.
+    }
+    return -1;
```

Comment:
I'm not a big fan of returning -1 here, I would like to see an std::optional<int> to force the user to check if it has the data they expect.
Thoughts anyone?

Reply:
Agree that we should try make use of `optional` here and "proper constructs" rather than sentinel values.

Reply:
Resolved